### PR TITLE
LibWeb+LibWasm: Some more JS API

### PR DESCRIPTION
**Pull request by alimpfard**
Currently impl'd:
- `WebAssembly.Table`

TODO:
- [x] Give some descriptions to those `RangeError` exceptions
- [x] ~~Add more boogs~~

### CODE REVIEW

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+JS::Value WebAssemblyTableConstructor::construct(FunctionObject&)
+{
+    auto& vm = this->vm();
+    auto& global_object = this->global_object();
+
+    auto descriptor = vm.argument(0).to_object(global_object);
+    if (vm.exception())
+        return {};
+
+    auto element_value = descriptor->get_without_side_effects("element");
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp:41]: `get_without_side_effects` is something you rarely need in general and pretty much never need in any "user-facing" code. As implemented, this wouldn't work with either Proxies or accessors. The spec is *really* unclear (`descriptor["element"]`? `:yakfused:`), but based on Firefox's implementation and a general git feeling this should be a regular `get()` + exception check :)

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    else if (element == "externref"sv)
+        reference_type = Wasm::ValueType(Wasm::ValueType::ExternReference);
+
+    if (!reference_type.has_value()) {
+        vm.throw_exception<JS::TypeError>(global_object, JS::ErrorType::InvalidHint, element);
+        return {};
+    }
+
+    auto initial_value = descriptor->get_without_side_effects("initial");
+    auto maximum_value = descriptor->get_without_side_effects("maximum");
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp:60]: Same as above

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        if (vm.exception())
+            return {};
+    }
+
+    if (maximum.has_value() && maximum.value() < initial) {
+        vm.throw_exception<JS::RangeError>(global_object, "maximum should be larger than or equal to initial");
+        return {};
+    }
+
+    auto value_value = descriptor->get_without_side_effects("value");
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp:84]: Same as above

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 4]
```diff
@@ -399,8 +416,22 @@ Optional<Wasm::Value> to_webassembly_value(JS::Value value, const Wasm::ValueTyp
         return Wasm::Value { static_cast<float>(number) };
     }
     case Wasm::ValueType::FunctionReference:
+    case Wasm::ValueType::NullFunctionReference: {
+        if (value.is_null())
+            return Wasm::Value { Wasm::ValueType(Wasm::ValueType::NullExternReference), 0ull };
+
+        if (value.is_function()) {
+            auto& fn = value.as_function();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:424]: This is not the serenity way

```suggestion
            auto& function = value.as_function();
```

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTablePrototype.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+namespace Web::Bindings {
+
+void WebAssemblyTablePrototype::initialize(JS::GlobalObject& global_object)
+{
+    Object::initialize(global_object);
+    define_native_accessor("length", length_getter, {}, JS::Attribute::Enumerable | JS::Attribute::Configurable);
+    define_native_function("grow", grow, 1, JS::Attribute::Writable | JS::Attribute::Enumerable | JS::Attribute::Configurable);
+    define_native_function("get", get, 1, JS::Attribute::Writable | JS::Attribute::Enumerable | JS::Attribute::Configurable);
+    define_native_function("set", set, 1, JS::Attribute::Writable | JS::Attribute::Enumerable | JS::Attribute::Configurable);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTablePrototype.cpp:19]:
```suggestion
    define_native_function("set", set, 2, JS::Attribute::Writable | JS::Attribute::Enumerable | JS::Attribute::Configurable);
```

**Review by linusg — COMMENTED** [review]
LGTM, couple of small tweaks

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp (hunk 1 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp:41]: I don't think it makes much sense for this to be anything except a string, but I'll go with your suggestion.

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTablePrototype.cpp (hunk 5 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTablePrototype.cpp:19]: `value` in `set()` is optional: https://webassembly.github.io/spec/js-api/#tables
`undefined set(unsigned long index, optional any value)`

Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTablePrototype.cpp:19]: Yet apparently Firefox sets the length to 2
This seems very inconsistent to me :thinking:

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp (hunk 1 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp:41]: > I don't think it makes much sense for this to be anything except a string

Sure, but that's not what I'm saying! This is about `new WebAssembly.Table({ get element() { ... } })` not working, for instance.

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTablePrototype.cpp (hunk 5 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTablePrototype.cpp:19]: cool, let's keep it like this then. Firefox boog spotted:

image

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        return {};
+    auto maximum_value = descriptor->get("maximum");
+    if (vm.exception())
+        return {};
+
+    if (initial_value.is_empty()) {
+        vm.throw_exception<JS::TypeError>(global_object, JS::ErrorType::NotA, "Number");
+        return {};
+    }
+
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp:72]: `get()` no longer returns empty values in non-error cases, so this never applies

```suggestion
```

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    auto initial = initial_value.to_u32(global_object);
+    if (vm.exception())
+        return {};
+
+    Optional<u32> maximum;
+
+    if (!maximum_value.is_empty()) {
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp:79]:
```suggestion
    if (!maximum_value.is_undefined()) {
```

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp (hunk 6 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyTableConstructor.cpp]: I stop looking at LibJS for like a month and then half the APIs have changed
neat :P

**Review by linusg — APPROVED** [review]

**Pull request merged**